docs/docs/validating-plugin-options.md

```diff
@@ -0,0 +1,230 @@
+- [Validate external access](#validate-external-access) where necessary
+- [Add custom error messages](#add-custom-error-messages) where useful
+- [Deprecate options](#deprecating-options) in a major version release rather than removing them
+
+### Add descriptions
+
+Make sure that every option and field has a `.description()` explaining its purpose. While this is not immediately visible, there might be tooling in the future that auto-generates plugin option documentation from the schema.
+
+### Set default options
+
+You can use the `.default()` method to default a value for an option. For example with `gatsby-plugin-console-log`, this is how it can default the `message` option to `"default message"` in case users do not pass their own `message`:
+
+```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
+exports.pluginOptionsSchema = ({ Joi }) => {
+  return Joi.object({
+    optionA: Joi.boolean().required().description(`Enables optionA.`),
+    message: Joi.string()
+      .default(`default message`) // highlight-line
+      .description(`The message logged to the console.`),
+    optionB: Joi.boolean().description(`Enables optionB.`),
+  })
+}
+```
+
+Accessing `pluginOptions.message` would then log `"default message"` in all plugin APIs if the user does not supply their own value:
+
+```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
+exports.onPreInit = (_, pluginOptions) => {
+  console.log(
+    `logging: "${pluginOptions.message}" to the console` // highlight-line
+  )
+}
+```
+
+### Validating external access
+
+Some plugins (particularly source plugins) query external APIs. With the `.external()` method, they can asynchronously validate that users have access to the API to provide a better experience if they pass invalid secrets.
+
+For example, this is how the [Contentful source plugin](/plugins/gatsby-source-contentful/) ensures the user has access to the space they are trying to query:
+
+```javascript:title=gatsby-source-contentful/gatsby-node.js
+exports.pluginOptionsSchema = ({ Joi }) => {
+  return Joi.object({
+    accessToken: Joi.string().required(),
+    spaceId: Joi.string().required(),
+    // ...more options here...
+  }).external(async pluginOptions => {
+    try {
+      await contentful
+        .createClient({
+          space: pluginOptions.spaceId,
+          accessToken: pluginOptions.accessToken,
+        })
+        .getSpace()
+    } catch (err) {
+      throw new Error(
+        `Cannot access Contentful space "${pluginOptions.spaceId}" with the provided access token. Double check they are correct and try again!`
+      )
+    }
+  })
+}
+```
+
+### Add custom error messages
+
+Sometimes you might want to provide more detailed error messages when validation fails for a specific field. Joi provides a [`.messages()` method](https://github.com/sideway/joi/issues/2109) which lets you override error messages for specific [error types](https://joi.dev/api/?v=17.2.1#list-of-errors).
```

Comment:
Why not link to https://joi.dev/api/?v=17.2.1#anymessagesmessages in first link?

Reply:
Because the documentation for that function is practically useless. 😬  It shows this:

<img width="530" alt="Screenshot 2020-10-20 at 08 33 24" src="https://user-images.githubusercontent.com/7525670/96548903-fb9e5800-12ae-11eb-8357-9ab9e7c893a5.png">

which links to

<img width="563" alt="Screenshot 2020-10-20 at 08 33 37" src="https://user-images.githubusercontent.com/7525670/96548917-00fba280-12af-11eb-9cd9-e8e61762a452.png">

neither of which tell the user anything about using `.messages()`. I'd rather link something that has a useful explanation that helps plugin authors dive deeper!

Reply:
But if the API gets changed the link to that issue is obsolete and we won't know it until someone runs into problems. You can add a note that it takes in the same object like `messages` on the `validate` function: https://joi.dev/api/?v=17.2.1#anyvalidatevalue-options